src/app/shared/clickOutSide/click-out-side.directive.ts

```diff
@@ -0,0 +1,45 @@
+import { DOCUMENT } from '@angular/common';
+import {
+  AfterViewInit,
+  Directive,
+  ElementRef,
+  EventEmitter,
+  Inject,
+  OnDestroy,
+  Output,
+} from '@angular/core';
+import { filter, fromEvent, Subscription } from 'rxjs';
+
+@Directive({
+  selector: '[appClickOutSide]',
+})
+export class ClickOutSideDirective implements AfterViewInit, OnDestroy {
+  @Output() clickOutside = new EventEmitter<void>();
+  documentClickSubscription: Subscription | undefined;
+
+  constructor(
+    private element: ElementRef,
+    @Inject(DOCUMENT) private document: Document
+  ) {}
+
+  ngOnDestroy(): void {
+    this.documentClickSubscription?.unsubscribe();
+  }
+
+  ngAfterViewInit(): void {
+    this.documentClickSubscription = fromEvent(this.document, 'click')
+      .pipe(
+        filter(() => {
+          return !this.isInside(event?.target as HTMLElement);
+        })
+      )
+      .subscribe(() => this.clickOutside.emit());
+  }
+
+  isInside(elementToCheck: HTMLElement): boolean {
+    return (
+      elementToCheck === this.element.nativeElement ||
+      this.element.nativeElement.contains(elementToCheck)
+    );
+  }
+}
```

Comment:
this directive is too complicated, use something like this based on HostListener
```
@Directive({
  selector: '[outsideClick]',
})
export class ClickOutsideDirective {
  @Output()
  outsideClick: EventEmitter<MouseEvent> = new EventEmitter();

  @HostListener('document:mousedown', ['$event'])
  onClick(event: MouseEvent): void {
    if (!this.elementRef.nativeElement.contains(event.target)) {
      this.outsideClick.emit(event);
    }
  }

  constructor(private elementRef: ElementRef) {}
}
```